File: backend/packages/carveo-core/src/carveo_core/valuation.py
from statistics import median
from typing import Literal

from carveo_core.contracts import DealPosition


def calculate_deal_position(price: int, comparable_prices: list[int]) -> DealPosition:
    if len(comparable_prices) < 3:
        return DealPosition(label="Limited data", sample_size=len(comparable_prices))

    typical_price = int(median(comparable_prices))
    difference_amount = price - typical_price
    difference_percent = round((difference_amount / typical_price) * 100, 2)
    if difference_percent <= -5:
        label: Literal["Below typical", "Near typical", "Above typical"] = "Below typical"
    elif difference_percent >= 5:
        label = "Above typical"
    else:
        label = "Near typical"
    return DealPosition(
        label=label,
        typical_price=typical_price,
        difference_amount=difference_amount,
        difference_percent=difference_percent,
        sample_size=len(comparable_prices),
    )

File: .pnpm-store/v11/projects/f4b7bd4ca3882799d979117dd8b23cc3/src/domain/valuation.ts
import type { DealPosition } from "@/domain/schemas";

export function median(values: number[]): number | null {
  if (values.length === 0) return null;
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? Math.round((sorted[middle - 1] + sorted[middle]) / 2) : sorted[middle];
}

export function calculateDealPosition(price: number, comparables: number[]): DealPosition {
  if (comparables.length < 4) {
    return { label: "Limited data", typicalPrice: null, differenceAmount: null, differencePercent: null, sampleSize: comparables.length };
  }
  const typicalPrice = median(comparables)!;
  const differenceAmount = price - typicalPrice;
  const differencePercent = Math.round((differenceAmount / typicalPrice) * 100);
  const label = differencePercent <= -8 ? "Below typical" : differencePercent >= 8 ? "Above typical" : "Near typical";
  return { label, typicalPrice, differenceAmount, differencePercent, sampleSize: comparables.length };
}
